### PR TITLE
[Artwork View] Auction flow buttons

### PR DESCRIPTION
**Pull request by kierangillen**
- Adds Auction bid buttons to Artwork view
- Moves `BidButton`, `BuyNowButton` and `MakeOfferButton` into children sub components of `CommercialButtons`

buttons_registrationcomplete

#skip_new_tests
- Added `#skip_new_tests` for rearranged button files which tests are covered in `CommercialButtons-tests` file

### CODE REVIEW

File: src/lib/Scenes/Artwork/Components/CommercialButtons/BidButton.tsx [hunk 1]
```diff
@@ -0,0 +1,153 @@
+import { Button, Sans } from "@artsy/palette"
+import { BidButton_artwork } from "__generated__/BidButton_artwork.graphql"
+import SwitchBoard from "lib/NativeModules/SwitchBoard"
+import React from "react"
+import { createFragmentContainer, graphql, RelayProp } from "react-relay"
+
+export const PREDICTION_URL = "https://live-staging.artsy.net"
```
Comment (ashfurrow) [inline src/lib/Scenes/Artwork/Components/CommercialButtons/BidButton.tsx]: We'll probably want this to be `live.artsy.net`. I've checked and we literally do the same thing for both:

https://github.com/artsy/eigen/blob/a677b504ccdb7fad552f7749912d2f048a925198/Artsy/App/ARSwitchBoard.m#L325-L326

But using the prod URL feels better.

File: src/lib/Scenes/Artwork/Components/CommercialButtons/BidButton.tsx [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+  relay: RelayProp
+}
+
+export class BidButton extends React.Component<BidButtonProps> {
+  setMaxBid = (newVal: number) => {
+    this.setState({ selectedMaxBidCents: newVal })
+  }
+
+  handleBid = () => {
+    SwitchBoard.presentModalViewController(this, `/bid/${this.props.artwork.slug}`)
```
Comment (ashfurrow) [inline src/lib/Scenes/Artwork/Components/CommercialButtons/BidButton.tsx]: I don't think this is the right URL format, it should be `/auction/:id/bid/:artwork_id`, right?

File: src/lib/Scenes/Artwork/Components/CommercialButtons/BidButton.tsx [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+  }
+
+  redirectToRegister = () => {
+    const { sale } = this.props.artwork
+    SwitchBoard.presentModalViewController(this, `/auction-registration/${sale.slug}`)
```
Comment (ashfurrow) [inline src/lib/Scenes/Artwork/Components/CommercialButtons/BidButton.tsx]: These `presentModalViewController` calls should be `presentNavigationViewController` (Eigen knows to present the right controllers modally, and to present the others as pushes.)

File: src/lib/Scenes/Artwork/Components/CommercialButtons/CommercialButtons.tsx [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+import React from "react"
+import { createFragmentContainer, graphql, RelayProp } from "react-relay"
+import { BidButtonFragmentContainer as BidButton } from "./BidButton"
+import { BuyNowButtonFragmentContainer as BuyNowButton } from "./BuyNowButton"
+import { MakeOfferButtonFragmentContainer as MakeOfferButton } from "./MakeOfferButton"
+
+export interface CommercialButtonProps {
+  artwork: CommercialButtons_artwork
+  relay: RelayProp
+  editionSetID?: string
```
Comment (ashfurrow) [inline src/lib/Scenes/Artwork/Components/CommercialButtons/CommercialButtons.tsx]: Can we add a comment here describing why we're using a prop for edition set ID, and not including it in the fragment? It makes sense to me now, but won't in six months.

File: src/lib/Scenes/Artwork/Components/CommercialButtons/CommercialButtons.tsx [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+export class CommercialButtons extends React.Component<CommercialButtonProps> {
+  handleInquiry = () => {
+    SwitchBoard.presentModalViewController(this, `/inquiry/${this.props.artwork.slug}`)
+  }
+
+  handleBid = () => {
+    SwitchBoard.presentModalViewController(this, `/bid/${this.props.artwork.slug}`)
```
Comment (ashfurrow) [inline src/lib/Scenes/Artwork/Components/CommercialButtons/CommercialButtons.tsx]: Same here with `presentNavigationViewController` vs `presentModalViewController`. Also I don't think this `handleBid` is getting called anywhere?

File: src/lib/Scenes/Artwork/Components/CommercialButtons/BidButton.tsx (hunk 2 above)
Comment (kierangillen) [reply src/lib/Scenes/Artwork/Components/CommercialButtons/BidButton.tsx]: Ah thanks

File: src/lib/Scenes/Artwork/Components/CommercialButtons/BidButton.tsx (hunk 3 above)
Comment (kierangillen) [reply src/lib/Scenes/Artwork/Components/CommercialButtons/BidButton.tsx]: Cool updating.

File: src/lib/Scenes/Artwork/Components/CommercialButtons/BidButton.tsx [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+      const myLastMaxBid = hasMyBids && myLotStanding.mostRecentBid.maxBid.cents
+      const increments = artwork.saleArtwork.increments.filter(increment => increment.cents > (myLastMaxBid || 0))
+      const firstIncrement = increments && increments.length && increments[0]
+      const incrementCents = firstIncrement && firstIncrement.cents
+
+      return (
+        <Button width={100} size="large" block onPress={() => this.redirectToBid(incrementCents)}>
+          {hasMyBids ? "Increase max bid" : "Bid"}
+        </Button>
+      )
```
Comment (ds300) [inline src/lib/Scenes/Artwork/Components/CommercialButtons/BidButton.tsx]: woah the logic in this method is hard to follow 😅 I don't know if it's possible to simplify so this is not a blocking comment. I'd be a bit scared to edit this code though haha.

File: src/lib/Scenes/Artwork/Components/CommercialButtons/CommercialButtons.tsx [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+export class CommercialButtons extends React.Component<CommercialButtonProps> {
+  handleInquiry = () => {
+    SwitchBoard.presentNavigationViewController(this, `/inquiry/${this.props.artwork.slug}`)
+  }
+
+  renderButtons = () => {
+    const { artwork } = this.props
+    const { isAcquireable, isOfferable, isInquireable, isInAuction } = artwork
+
+    if (isInAuction && artwork.sale && !artwork.sale.isClosed) {
```
Comment (ds300) [inline src/lib/Scenes/Artwork/Components/CommercialButtons/CommercialButtons.tsx]: not a big deal but these two latter conditions are already taken care of in the bid button render method, so you could remove these checks and also the sale.isClosed selection from this fragment.

File: src/lib/Scenes/Artwork/Components/CommercialButtons/CommercialButtons.tsx [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+          Contact gallery
+        </Button>
+      )
+    } else {
+      return <></>
+    }
+  }
+
+  render() {
+    return <>{this.renderButtons()}</>
```
Comment (ds300) [inline src/lib/Scenes/Artwork/Components/CommercialButtons/CommercialButtons.tsx]: haha this indirection seems a bit useless now 🙈

File: src/lib/Scenes/Artwork/Components/CommercialButtons/__tests__/BidButton-tests.tsx [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+
+    it("with bidder registration pending approval", async () => {
+      const artwork = merge({}, ArtworkFromLiveAuctionRegistrationOpen, BidderPendingApproval)
+      const wrapper = await getWrapper(artwork)
+
+      expect(wrapper.find(Button).text()).toContain("Enter live bidding")
+    })
+
+    it("with registered bidder", async () => {
+      const artwork = merge({}, ArtworkFromLiveAuctionRegistrationOpen, RegisteredBidder)
```
Comment (ds300) [inline src/lib/Scenes/Artwork/Components/CommercialButtons/__tests__/BidButton-tests.tsx]: #nonblocking why use lodash's merge instead of `{..spreading}`?

File: src/lib/Scenes/Artwork/Components/CommercialButtons/__tests__/CommercialButtons-tests.tsx [hunk 10]
```diff
@@ -43,76 +34,78 @@ const componentWithQuery = async ({ mockArtworkData, mockOrderMutationResults, m
...
-      <CommercialButtons
-        relay={{ environment: {} } as RelayProp}
-        artwork={artwork as CommercialButtonProps["artwork"]}
-      />
-    )
-    expect(component.text()).toContain("Contact gallery")
+    const commercialButtons = await relayComponent({
+      artwork,
+    })
+    expect(commercialButtons.text()).toContain("Contact gallery")
```
Comment (ds300) [inline src/lib/Scenes/Artwork/Components/CommercialButtons/__tests__/CommercialButtons-tests.tsx]: This test case appears to not match the description. looks like the description is correct but the test case is a duplication of the one above.

File: src/lib/Scenes/Artwork/Components/CommercialButtons/__tests__/CommercialButtons-tests.tsx [hunk 11]
```diff
@@ -43,76 +34,78 @@ const componentWithQuery = async ({ mockArtworkData, mockOrderMutationResults, m
   })
 }
 
+const relayComponent = async ({ artwork }) => {
```
Comment (ds300) [inline src/lib/Scenes/Artwork/Components/CommercialButtons/__tests__/CommercialButtons-tests.tsx]: would it make sense to add a test for the bid button in here?

File: src/lib/Scenes/Artwork/Components/CommercialButtons/CommercialButtons.tsx (hunk 8 above)
Comment (kierangillen) [reply src/lib/Scenes/Artwork/Components/CommercialButtons/CommercialButtons.tsx]: Agreed!

File: src/lib/Scenes/Artwork/Components/CommercialButtons/BidButton.tsx (hunk 6 above)
Comment (kierangillen) [reply src/lib/Scenes/Artwork/Components/CommercialButtons/BidButton.tsx]: This was copied over from Reaction so gonna leave it as it reflects working logic.

File: src/lib/Scenes/Artwork/Components/CommercialButtons/__tests__/CommercialButtons-tests.tsx (hunk 11 above)
Comment (kierangillen) [reply src/lib/Scenes/Artwork/Components/CommercialButtons/__tests__/CommercialButtons-tests.tsx]: Added!

File: src/lib/Scenes/Artwork/Components/CommercialButtons/__tests__/CommercialButtons-tests.tsx (hunk 10 above)
Comment (kierangillen) [reply src/lib/Scenes/Artwork/Components/CommercialButtons/__tests__/CommercialButtons-tests.tsx]: Fixed this.

File: src/lib/Scenes/Artwork/Components/CommercialButtons/__tests__/BidButton-tests.tsx (hunk 9 above)
Comment (kierangillen) [reply src/lib/Scenes/Artwork/Components/CommercialButtons/__tests__/BidButton-tests.tsx]: Going to keep as is for now.

**Comment by ds300** [comment]
#mergeOnGreen
